packages/flutter_tools/lib/src/devfs.dart

```diff
@@ -205,107 +301,130 @@ class DevFS {
         if (uri.toString() == 'lib/')
           continue;
         Directory directory = new Directory.fromUri(uri);
-        printTrace('DevFS: Syncing package $packageName started');
-        bool packageWritten =
-            await _syncDirectory(directory,
+        bool packageExists =
+            await _scanDirectory(directory,
                                  directoryName: 'packages/$packageName',
                                  recursive: true);
-        if (packageWritten) {
+        if (packageExists) {
           sb ??= new StringBuffer();
           sb.writeln('$packageName:packages/$packageName');
         }
-        printTrace('DevFS: Syncing package $packageName finished');
       }
     }
+    status.stop(showElapsedTime: true);
     if (bundle != null) {
+      status = logger.startProgress('Scanning asset files...');
       // Synchronize asset bundle.
       for (AssetBundleEntry entry in bundle.entries) {
         // We write the assets into 'build/flx' so that they are in the
         // same location in DevFS and the iOS simulator.
         final String devicePath = path.join('build/flx', entry.archivePath);
-        if (!bundleDirty && entry.isStringEntry) {
-          // When the bundle isn't dirty, we do not need to sync string
-          // entries.
-          continue;
-        }
-        _syncBundleEntry(devicePath, entry);
+        _scanBundleEntry(devicePath, entry, bundleDirty);
       }
+      status.stop(showElapsedTime: true);
     }
     // Handle deletions.
+    status = logger.startProgress('Scanning for deleted files...');
     final List<String> toRemove = new List<String>();
     _entries.forEach((String path, DevFSEntry entry) {
-      if (!entry._wasSeen) {
-        _deleteEntry(path, entry);
+      if (!entry._exists) {
+        _deletedEntries.add(entry);
         toRemove.add(path);
       }
     });
     for (int i = 0; i < toRemove.length; i++) {
       _entries.remove(toRemove[i]);
     }
-    // Send the assets.
-    printTrace('DevFS: Waiting for sync of ${_pendingWrites.length} files '
-               'to finish');
+    status.stop(showElapsedTime: true);
+
+    if (_deletedEntries.length > 0) {
+      status = logger.startProgress('Removing deleted files...');
+      for (DevFSEntry entry in _deletedEntries) {
+        print(entry.devicePath);
+        Future<Response> operation = _operations.deleteFile(fsName, entry);
+        if (operation != null)
+          _pendingOperations.add(operation);
+      }
+      await Future.wait(_pendingOperations);
+      _pendingOperations.clear();
+      _deletedEntries.clear();
+      status.stop(showElapsedTime: true);
+    } else {
+      printStatus("No files to remove.");
```

Comment:
We probably don't need to print this out each time - there generally won't be deleted files.